test/behaviour/resolution/framework/common/RuleResolutionBuilder.java

```diff
@@ -119,34 +125,52 @@
         return Graql.and(result);
     }
 
-    public LinkedHashMap<String, Statement> statementToResolutionProperties(Statement statement) {
+    public LinkedHashMap<String, Statement> statementToResolutionProperties(final Transaction tx, Statement statement, final Boolean inferred) {
         LinkedHashMap<String, Statement> props = new LinkedHashMap<>();
 
-        String statementVar = statement.var().name();
+        String statementVarName = statement.var().name();
 
         for (VarProperty varProp : statement.properties()) {
 
             if (varProp instanceof HasAttributeProperty) {
                 String nextVar = getNextVar("x");
-                StatementInstance propStatement = Graql.var(nextVar).isa("has-attribute-property").has((HasAttributeProperty) varProp).rel("owner", statementVar);
+                StatementInstance propStatement = Graql.var(nextVar).isa("has-attribute-property")
+                        .rel("owned", ((HasAttributeProperty) varProp).attribute().var().name())
+                        .rel("owner", statementVarName);
+                if (inferred != null) {
+                    propStatement = propStatement.has("inferred", inferred);
+                }
                 props.put(nextVar, propStatement);
 
-            } else if (varProp instanceof RelationProperty){
+            } else if (varProp instanceof RelationProperty) {
                 for (RelationProperty.RolePlayer rolePlayer : ((RelationProperty)varProp).relationPlayers()) {
                     Optional<Statement> role = rolePlayer.getRole();
 
                     String nextVar = getNextVar("x");
 
-                    StatementInstance propStatement = Graql.var(nextVar).isa("relation-property").rel("rel", statementVar).rel("roleplayer", Graql.var(rolePlayer.getPlayer().var()));
-                    if(role.isPresent()) {
+                    StatementInstance propStatement = Graql.var(nextVar).isa("relation-property").rel("rel", statementVarName).rel("roleplayer", Graql.var(rolePlayer.getPlayer().var()));
+                    if (role.isPresent()) {
                         String roleLabel = ((TypeProperty) getOnlyElement(role.get().properties())).name();
-                        propStatement = propStatement.has("role-label", roleLabel);
+                        final Set<Role> roles = tx.getRole(roleLabel).sups().collect(Collectors.toSet());
```

Comment:
we might be able to use a .forEach loop here? not sure because we modify state though hummm. might be best approach already

Reply:
Yeah I think I prefer the for loop because of the state modification.